---
title: 'Class 15: Transcriptomics and RNA Seq Analysis'
author: 'Jasmine Lee (PID: A15583527)'
date: "11/16/2021"
output: pdf_document
---
# Background

Our data for today comes from Himes et al. RNASeq analysis of the drug dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

Read the countData and colData.

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <- read.csv("airway_metadata.csv")
```

Now take a look at each.

```{r}
head(counts)
head(metadata)
```

> **Q1**. How many genes are in this dataset?

```{r}
nrow(counts)
```

*Answer*: There are 38,694 genes in this dataset.


> **Q2**. How many 'control' cell lines do we have?

```{r}
sum(metadata$dex == "control")
```

*Answer*: There are 4 'control' cell lines.


# Toy Differential Gene Expression

First I need to extract all the 'control' columns. THen I will take the row-wise mean to get the average count values for all genes in these four experiments.

```{r}
control.inds <- metadata$dex=="control"
control.counts <- counts[ ,control.inds]
head(control.counts)
control.mean <- rowMeans(control.counts)
```


> **Q3**. How would you make the above code (in the hands-on worksheet) in either approach more robust?

*Answer*: I would make the above code in either approach shown in the hands-on section worksheet more robust by using the rowMeans function instead of finding the mean by dividing rowSums by the number of rows. This would allow for me to easily use this code on datasets with more or less samples.


> **Q4**. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean).

*Answer*:

Now do the same for all the 'treated' columns.

```{r}
treated.inds <- metadata$dex=="treated"
treated.counts <- counts[ ,treated.inds]
head(treated.counts)
treated.mean <- rowMeans(treated.counts)
```


We will combine our meancount data for bookkeeping purposes.

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```


> **Q5(a)**. Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

*Answer*:

Let's make a quick plot.

```{r}
plot(meancounts)
```

> **Q5(b)**. You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot?

*Answer*: I would use the geom_point() function for this plot.

```{r}
library(ggplot2)
ggplot(meancounts) +
  aes(control.mean,treated.mean) +
  geom_point()
```


> **Q6**. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

*Answer*: The argument to plot() is log.

```{r}
plot(meancounts, log="xy")
```

We often use log2 in this field because it has nice math properties that make interpretation easier.

```{r}
# No change in gene expression
log2(10/10)
```
```{r}
# Doubled increase in gene expression
log2(20/10)
```
```{r}
log2(40/10)
```
```{r}
log2(5/10)
```

Cool, we see 0 values for no change, + values for upregulation, and - values for decreases. This nice property leads us to work with **log2(fold-change)** all the time in the genomics and proteomics field.

Let's add the **log2(fold-change)** values to our 'meancounts' data.

```{r}
meancounts$log2fc <- log2(meancounts[ ,"treated.mean"]/
                            meancounts[ ,"control.mean"])
head(meancounts)
```

NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. Let's filter our data to remove these genes because we can't say anything about these as we have no data for them.

```{r}
zero.vals <- which(meancounts[ ,1:2]==0, arr.ind=TRUE)
to.rm <- unique(zero.vals[ ,1])
head(zero.vals)
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

How many do we have left?

```{r}
nrow(mycounts)
```

> **Q7**. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

*Answer*: The arr.ind argument provides the columns and rows for any object in the array that is TRUE for being equal to 0. The purpose of taking the first column of the output and calling the unique() function is to make sure that rows that have two zeroes are not counted twice.


```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```


> **Q8**. Using the up.ind vector above, can you determine how many upregulated genes we have at the greater than 2 fc level?

```{r}
sum(up.ind)
```

*Answer*: There are 250 upregulated genes at the greater than 2 fc level.


> **Q9**. Using the down.ind vector above, can you determine how many downregulated genes we have at the less than -2 fc level?

```{r}
sum(down.ind)
```

*Answer*: There are 367 downregulated genes at the less than -2 fc level.


> **Q10**. Do you trust these results? Why or why not?

*Answer*: No, I do not trust these results. We have not normalized the fold changes or analyzed them to determine statistical significance yet.


# DESeq2 Analysis

Let’s do this the right way. DESeq2 is an R package specifically for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor.

```{r}
library(DESeq2)
```

We will use the DESeqDataSetFromMatrix() function to built the required DESeqDataSet object and call it dds.

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts,
                              colData=metadata, 
                              design=~dex)
dds
```

Now we can run DESeq analysis.

```{r}
dds <- DESeq(dds)
```

To get the results here, we use the DESeq 'results()' function:

```{r}
res <- results(dds)
res
```


# Save Our Results

Write out whole results dataset (including genes that don't change significantly).

```{r}
write.csv(res, file="allmyresults.csv")
```

Focus in on those genes with a small p-value (i.e. show a significant change).

```{r}
res05 <- results(dds, alpha=0.05)
```

```{r}
summary(res05)
```


# Volcano Plots

Let’s make a commonly produced visualization from this data, namely a so-called Volcano plot. These summary figures are frequently used to highlight the proportion of genes that are both significantly regulated and display a high fold change.

```{r}
plot(res$log2FoldChange, res$padj)
```

That is not a useful plot because all the small p-values are hidden at the bottom of the plot, and we can't really see them. log will help.

```{r}
plot(res$log2FoldChange, log(res$padj))
```

```{r}
plot(res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
```

Finally let's add some color to this plot to draw attention to the genes (i.e. points) we care about-- that is those with large fold-change and low p-values (i.e. high -log(pvalues)).

```{r}
plot( res$log2FoldChange,  -log(res$padj), 
 ylab="-Log(P-value)", xlab="Log2(FoldChange)")

# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
```

We will now add some color to the points:

```{r}
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[abs(res$log2FoldChange) > 2]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2)
mycols[inds] <- "blue"

# Volcano plot with custom colors 
plot(res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)")

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```


# Adding Annotation Data

For this, we need two Bioconductor packages.

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```

Let's have a loook at what is in the 'org.Hs.eg.db'.

```{r}
columns(org.Hs.eg.db)
```

We will use the 'mapIDs' function to translate between identifiers from different databases.

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",   # Their format
                     column="SYMBOL",     # New format we want
                     multiVals="first")
```

```{r}
head(res)
```


> **Q11**. Run the **mapIds()* function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot, and res$genename.

*Answer*:

We need ENTREZ ids for pathway analysis with KEGG.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",   # Their format
                     column="ENTREZID",     # New format we want
                     multiVals="first")
```

```{r}
res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",   # Their format
                     column="UNIPROT",     # New format we want
                     multiVals="first")
```

```{r}
res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",   # Their format
                     column="GENENAME",     # New format we want
                     multiVals="first")
```


Let's make another volcano plot with some gene labels.
For this, we can use the **EnhancedVolcano** package.

```{r}
library(EnhancedVolcano)
```

```{r}
x <- as.data.frame(res)

EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```


# Pathway Analysis/Geneset Annotation

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```

The main **gage()** function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

```{r}
foldchanges <- res$log2FoldChange
names(foldchanges) <- res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

This separates out results by "greater" and "less" (i.e. those that are upregulated and those that are downregulated).

```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```

Now, let’s try out the 'pathview()' function from the **pathview** package to make a pathway plot with our RNA-Seq expression results shown in color.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```


![](hsa05310.pathview.png)


> **Q12**. Can you do the same procedure as above to plot the pathview figures for the top 2 downregulated pathways?

*Answer*: 

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05332")
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04940")
```

![](hsa05332.pathview.png)

![](hsa04940.pathview.png)
