---
title: 'Class 12: Structural Bioinformatics II'
author: 'Jasmine Lee (PID: A15583527)'
date: "11/4/2021"
output:
  word_document: default
  pdf_document: default
---


# 3. Using Bio3D

I need to load the library.
```{r}
library(bio3d)

pdb <- read.pdb("1hsg")
pdb
```

```{r}
aa123(pdbseq(pdb))
```

Plot of B-factor

```{r}
plot.bio3d(pdb$atom$b, sse=pdb)
```

The ATOM records

```{r}
head(pdb$atom)
```

> **Q7**. How many amino acid residues are there in this pdb object?

*Answer*: There are 198 amino acid residues.


> **Q8**. Name one of the two non-protein residues.

*Answer*: One of the two non-protein residues is HOH.


> **Q9**. How many protein chains are in this structure?

*Answer*: There are 2 protein chains in this structure.


# 4. Comparative Analysis of Adenylate Kinase

```{r}
# install.packages("bio3d")
# install.packages("ggplot2")
# install.packages("ggrepel")
# install.packages("devtools")
# install.packages("BiocManager")

# BiocManager::install("msa")
# devtools::install_bitbucket("Grantlab/bio3d-view")
```


> **Q10**. Which of the packages above is found only on BioConductor and not CRAN?

*Answer*: The "msa" package is found only on BioConductor.


> **Q11**. Which of the above packages is not found on Bioconductor or CRAN?

*Answer*: The "bio3d-view" package is not found on BioConductor or CRAN.


> **Q12**. True or false? Functions from the devtools package can be used to install packages from GitHub and BitBucket.

*Answer*: True.


```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

> **Q13**. How many amino acids are in this sequence, i.e. how long is this sequence?

*Answer*: There are 214 amino acids in this sequence.


# Comparative Analysis of Protein Structures

Using the Bio3D package
```{r}
library(bio3d)

pdb <- read.pdb("1hel")
pdb
```

Let's use a bioinformatics method called NMA (Normal Mode Analysis) to predict the dynamics (flexibility) of this enzyme.

```{r}
modes <- nma(pdb)
plot(modes)
```

Make a "movie" of its predicted motion. We often call this a "trajectory".

```{r}
mktrj(modes, file="nma.pdb")
```


# VMD Visualization Image

![](vmdmotion.png)


# Analysis of ADK

```{r}
aa <- get.seq("1ake_A")
aa
```

```{r}
# Run BLAST from R
blast <- blast.pdb(aa)
```

```{r}
hits <- plot(blast)
```

```{r}
hits$pdb.id
```

```{r}
# Download related PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

Multiple Structure Alignment
```{r}
pdbs <- pdbaln(files, fit=TRUE)
```

```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels=ids)
```


# Viewing our superposed structures

```{r}
library(bio3d.view)
library(rgl)

view.pdbs(pdbs)
```


# Principal Component Analysis

We will use the bio3d pca() function, which is designed for protein structure data.

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```


# Optional Further Visualization

Make a trajectory visualization of the motion captured by the first Prinicpal Component.

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```


# VMD Visualization Image

![](vmdtrajectory.png)


# Find-a-Gene VMD Visualization Image

![](p53vmdscene.png)


# Normal Mode Analysis

```{r}
# NMA of all structures
modes <- nma(pdbs)
plot(modes, pdbs, col=grps.rd)
```

> **Q14**. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

*Answer*: In this plot, the colored lines are similar to each other while the black lines are similar to each other. However, the black and colored lines are different when compared. They differ the most in residues 1-60 and residues 110-190. These regions are likely responsible for nucleotide binding and change conformation in bound and unbound states.